Uploading file with offset uploads the whole file

When passing an open file that has been seeked in to `PutObject()`, an error occurs because it thinks it needs to upload the whole file. The following program shows it (maybe needs the code from #678):

```go
package main

import (
	crand "crypto/rand"
	"fmt"
	"io"
	"io/ioutil"
	"math/rand"
	"os"
	"time"

	minio "github.com/minio/minio-go"
)

func main() {
	rand.Seed(time.Now().UnixNano())

	var length = 120000
	data := make([]byte, length)
	_, err := io.ReadFull(crand.Reader, data)
	if err != nil {
		panic(err)
	}

	tempfile, err := ioutil.TempFile("", "minio-go-upload-test-")
	if err != nil {
		panic(err)
	}

	if err = tempfile.Close(); err != nil {
		panic(err)
	}

	if err = ioutil.WriteFile(tempfile.Name(), data, 0600); err != nil {
		panic(err)
	}

	tempfile, err = os.Open(tempfile.Name())
	if err != nil {
		panic(err)
	}

	client, err := minio.New("s3.amazonaws.com", os.Getenv("AWS_ACCESS_KEY_ID"), os.Getenv("AWS_SECRET_ACCESS_KEY"), false)
	if err != nil {
		panic(err)
	}

	// client.TraceOn(os.Stderr)

	name := fmt.Sprintf("test-file-%v", rand.Uint32())

	offset := 123
	if _, err := tempfile.Seek(int64(offset), io.SeekStart); err != nil {
		panic(err)
	}

	n, err := client.PutObject(os.Getenv("BUCKET"), name, tempfile, "binary/octet-stream")
	fmt.Printf("wrote %d bytes to %v err %v\n", n, name, err)

	if n != int64(length-offset) {
		panic(fmt.Sprintf("invalid length returned, want %v, got %v", int64(length-offset), n))
	}
}
```

```
$ BUCKET=restic-test-travis go run main.go
wrote 0 bytes to test-file-1876261662 err Data read ‘119877’ is not equal to the size ‘120000’ of the input Reader.
panic: invalid length returned, want 119877, got 0

goroutine 1 [running]:
main.main()
	/home/fd0/tmp/main.go:61 +0x775
exit status 2
```

--- a/api-put-object.go
+++ b/api-put-object.go
@@ -109,14 +109,24 @@ func getReaderSize(reader io.Reader) (size int64, err error) {
 			case "|0", "|1":
 				return
 			}
-			size = st.Size()
+			var pos int64
+			pos, err = v.Seek(0, 1) // SeekCurrent.
+			if err != nil {
+				return -1, err
+			}
+			size = st.Size() - pos
 		case *Object:
 			var st ObjectInfo
 			st, err = v.Stat()
 			if err != nil {
 				return
 			}
-			size = st.Size
+			var pos int64
+			pos, err = v.Seek(0, 1) // SeekCurrent.
+			if err != nil {
+				return -1, err
+			}
+			size = st.Size - pos
 		}
 	}
 	// Returns the size here.

--- a/api_functional_v4_test.go
+++ b/api_functional_v4_test.go
@@ -2477,3 +2477,97 @@ func TestGetObjectObjectModified(t *testing.T) {
 		t.Errorf("Expected ReadAt to fail with error %s but received %s", s3ErrorResponseMap["PreconditionFailed"], err.Error())
 	}
 }
+
+// Test validates putObject to upload a file seeked at a given offset.
+func TestPutObjectUploadSeekedObject(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping functional tests for the short runs")
+	}
+
+	// Instantiate new minio client object.
+	c, err := NewV4(
+		os.Getenv("S3_ADDRESS"),
+		os.Getenv("ACCESS_KEY"),
+		os.Getenv("SECRET_KEY"),
+		mustParseBool(os.Getenv("S3_SECURE")),
+	)
+	if err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	// Enable tracing, write to stderr.
+	// c.TraceOn(os.Stderr)
+
+	// Set user agent.
+	c.SetAppInfo("Minio-go-FunctionalTest", "0.1.0")
+
+	// Make a new bucket.
+	bucketName := randString(60, rand.NewSource(time.Now().UnixNano()), "minio-go-test")
+	err = c.MakeBucket(bucketName, "us-east-1")
+	if err != nil {
+		t.Fatal("Error:", err, bucketName)
+	}
+	defer c.RemoveBucket(bucketName)
+
+	tempfile, err := ioutil.TempFile("", "minio-go-upload-test-")
+	if err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	var length = 120000
+	data := bytes.Repeat([]byte("1"), length)
+
+	if _, err = tempfile.Write(data); err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	objectName := fmt.Sprintf("test-file-%v", rand.Uint32())
+
+	offset := length / 2
+	if _, err := tempfile.Seek(int64(offset), 0); err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	n, err := c.PutObject(bucketName, objectName, tempfile, "binary/octet-stream")
+	if err != nil {
+		t.Fatal("Error:", err)
+	}
+	if n != int64(length-offset) {
+		t.Fatalf("Invalid length returned, want %v, got %v", int64(length-offset), n)
+	}
+	tempfile.Close()
+	if err = os.Remove(tempfile.Name()); err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	length = int(n)
+
+	obj, err := c.GetObject(bucketName, objectName)
+	if err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	n, err = obj.Seek(int64(offset), 0)
+	if err != nil {
+		t.Fatal("Error:", err)
+	}
+	if n != int64(offset) {
+		t.Fatalf("Invalid offset returned, want %v, got %v", int64(offset), n)
+	}
+
+	n, err = c.PutObject(bucketName, objectName+"getobject", obj, "binary/octet-stream")
+	if err != nil {
+		t.Fatal("Error:", err)
+	}
+	if n != int64(length-offset) {
+		t.Fatalf("Invalid length returned, want %v, got %v", int64(length-offset), n)
+	}
+
+	if err = c.RemoveObject(bucketName, objectName); err != nil {
+		t.Fatal("Error:", err)
+	}
+
+	if err = c.RemoveObject(bucketName, objectName+"getobject"); err != nil {
+		t.Fatal("Error:", err)
+	}
+}

